### ISSUE TITLE
xgboost and nrounds as budget

### ISSUE THREAD

**Issue by berndbischl** [mlr-org/mlr3hyperband#18]
it seems to be that we can only handle numerical params as budget? if so, also this needs to be documented and asserted

but this seems quite restrictive. what do we do now? use a trafo?

**Comment by berndbischl** [comment]
this seems to work out-of-the-box

```
ps = ParamSet$new(list(
  ParamInt$new("nrounds", lower = 1, upper = 100, tags = "budget"),
  ParamDbl$new("eta", lower = 0, upper = 1)
))
ps$trafo = function(x, param_set) {
  x$nrounds = round(x$nrounds)
  return(x)
}
```


as long as this "trick" is documented for the user (and the doc issues from my original post are done) i think i am ok with this

**Comment by berndbischl** [comment]
actually, we need a unit test for this and at least a mini-change 

```
ps = ParamSet$new(list(
  ParamInt$new("nrounds", lower = 1, upper = 100, tags = "budget"),
  ParamDbl$new("eta", lower = 0, upper = 1)
))
ps$trafo = function(x, param_set) {
  x$nrounds = round(x$nrounds)
  return(x)
}


inst = TuningInstance$new(
  tsk("iris"),
  lrn("classif.xgboost"),
  rsmp("holdout"),
  msr("classif.ce"),
  ps,
  term("evals", n_evals = 100000)
)


tuner = TunerHyperband$new(eta = 2L, use_subsamp = FALSE)
tuner$tune(inst)
```

this produces an error

```
INFO  [13:31:41.302] Finished tuning after 301 evals
Error in instance$assign_result(pv, perf) :
  Assertion on 'tune_x' failed: nrounds: Must be of type 'single i
ntegerish value', not 'double'.
```

**Comment by SebGruber1996** [comment]
The error seems to be a bug in `mlr3tuning`.
First of all, all evaluation have finished and if you check the nrounds with the following line
`as.data.frame(inst$archive()[, .(nrounds = sapply(params, "[", "nrounds")
)])`, you can see that all of them are integerish.
Additionally the error will not appear if upper = 10 for nrounds, but the calculated budgets are still doubles.
As a last note, your total eval amount is 301, but I got 445 under the same settings.
I will document the transformer, but I dont think I can fix the error in mlr3hyperband

**Comment by SebGruber1996** [comment]
ok, this is getting awkward...
this fails with the same error:
```
  set.seed(123)
   
  # define hyperparameter and budget parameter for tuning with hyperband
  ps = ParamSet$new(list(
    ParamInt$new("nrounds", lower = 1, upper = 10, tags = "budget"),
    ParamDbl$new("eta",     lower = 0, upper = 1),
    ParamFct$new("booster", levels = c("gbtree", "gblinear", "dart"))
  ))

  ps$trafo = function(x, param_set) {
    x$nrounds = round(x$nrounds)
    return(x)
  }

  inst = TuningInstance$new(
    tsk("iris"),
    lrn("classif.xgboost"),
    rsmp("holdout"),
    msr("classif.ce"),
    ps,
    term("evals", n_evals = 100000)
  )

  # hyperband + tuning
  tuner = TunerHyperband$new(eta = 1.9)
  expect_tuner(tuner)
  tuner$tune(inst)
```

but removing the !!!eta!!! parameter causes the error to disappear (even though the error clearly blames "nrounds" as the evil guy):
```
  set.seed(123)
   
  # define hyperparameter and budget parameter for tuning with hyperband
  ps = ParamSet$new(list(
    ParamInt$new("nrounds", lower = 1, upper = 10, tags = "budget"),
    ParamFct$new("booster", levels = c("gbtree", "gblinear", "dart"))
  ))

  ps$trafo = function(x, param_set) {
    x$nrounds = round(x$nrounds)
    return(x)
  }

  inst = TuningInstance$new(
    tsk("iris"),
    lrn("classif.xgboost"),
    rsmp("holdout"),
    msr("classif.ce"),
    ps,
    term("evals", n_evals = 100000)
  )

  # hyperband + tuning
  tuner = TunerHyperband$new(eta = 1.9)
  expect_tuner(tuner)
  tuner$tune(inst)
```

**Comment by berndbischl** [comment]
i have disabled the unit test for the integer trafo now, as it breaks travis. 
it must be reeanabled before this is closed

**Comment by berndbischl** [comment]
#test_that("TunerHyperband using param trafo and non-integer eta", {

#  set.seed(123)
   
#  # define hyperparameter and budget parameter for tuning with hyperband
#  ps = ParamSet$new(list(
#    ParamInt$new("nrounds", lower = 1, upper = 10, tags = "budget"),
#    #ParamDbl$new("eta",     lower = 0, upper = 1),
#    ParamFct$new("booster", levels = c("gbtree", "gblinear", "dart"))
#  ))

**Comment by mb706** [comment]
You should probably round on [this line](https://github.com/mlr-org/mlr3hyperband/blob/4ea7a505069d8a9320ca6cc6f707912b3b0e362f/R/TunerHyperband.R#L397) if the budget paramter is a ParamInt

**Comment by SebGruber1996** [comment]
ok, I got it now.
The new fix even removes the requirement of a user-defined parameter trafo.
I also readded the unit test.
Example (similar to the unit test):
```
set.seed(123)
   
ps = ParamSet$new(list(
  ParamInt$new("nrounds", lower = 1, upper = 10, tags = "budget"),
  ParamDbl$new("eta",     lower = 0, upper = 1),
  ParamFct$new("booster", levels = c("gbtree", "gblinear", "dart"))
))

# not required anymore
#ps$trafo = function(x, param_set) {
#  x$nrounds = round(x$nrounds)
#  return(x)
#}

inst = TuningInstance$new(
  tsk("iris"),
  lrn("classif.xgboost"),
  rsmp("holdout"),
  msr("classif.ce"),
  ps,
  term("evals", n_evals = 100000)
)

# hyperband + tuning
tuner = TunerHyperband$new(eta = 3.9)
tuner$tune(inst)

results = inst$archive()[, .(
  nrounds = sapply(params, "[", "nrounds"),
  eta     = sapply(params, "[", "eta"),
  booster = sapply(params, "[", "booster"),
  classif.ce
)]
```
